### ISSUE TITLE
Allow using only rootless podman without errors

### ISSUE THREAD

**Issue by MartinX3** [cockpit-project/cockpit-podman#742]
I only use the user podman.service, user podman.socket and user podman-auto-update service.

So I disabled the rootful service,socket, podman-auto-updater service.

Problems
- I still get the "the podman system service is also available" message.
- The "services" tab shows me a red ! about the podman service not running.
- In the "services" tab, I can't activate the "user" podman-auto-update service or look at its status.
Only the rootful one is listet. (Same for the podman service)

**Comment by KKoukiou** [comment]
Hey @MartinX3,

> I only use the user podman.service, user podman.socket and user podman-auto-update service.
> 
> So I disabled the rootful service,socket, podman-auto-updater service.
> 
> Problems
> 
>     * I still get the "System Podman service is also available" message.

This is desired and designed this way. It's an info notification and you can always dismiss it.

> 
>     * The "services" tab shows me a red ! about the podman service not running.

This is also expected, systemctl list-units will also show this unit as failed, and we don't do anything smarter than showing all units here, as provided myb the API.

> 
>     * In the "services" tab, I can't activate the "user" podman-auto-update service or look at its status.
>       Only the rootful one is listet. (Same for the podman service)
> 

User services are not supported yet, there is an issue about it https://github.com/cockpit-project/cockpit/issues/13217

**Comment by MartinX3** [comment]
Hey @KKoukiou,
thank you for your response.

> This is desired and designed this way. It's an info notification and you can always dismiss it.

It would be nice, if I could deactivate this info message.

> > ```
> > * The "services" tab shows me a red ! about the podman service not running.
> > ```
> 
> This is also expected, systemctl list-units will also show this unit as failed, and we don't do anything smarter than showing all units here, as provided myb the API.

It would be nice, if there is a blacklist to filter units I want to fail.
So I won't get notified about them.

> User services are not supported yet, there is an issue about it [cockpit-project/cockpit#13217](https://github.com/cockpit-project/cockpit/issues/13217)

Thank you for this issue ticket, I'll subscribe it.

**Comment by KKoukiou** [comment]
> Hey @KKoukiou,
> thank you for your response.
> 
> > This is desired and designed this way. It's an info notification and you can always dismiss it.
> 
> It would be nice, if I could deactivate this info message.

Alright, we can store this information, and if it's dismissed one time, then one next session it won't appear again.

> 
> > This is also expected, systemctl list-units will also show this unit as failed, and we don't do anything smarter than showing all units here, as provided myb the API.
> 
> It would be nice, if there is a blacklist to filter units I want to fail.
> So I won't get notified about them.

I find this the wrong approach to deal with your issue If you want this unit to not start just disable it. systemctl disable podman. systemctl reset-failed. 

> 
> > User services are not supported yet, there is an issue about it [cockpit-project/cockpit#13217](https://github.com/cockpit-project/cockpit/issues/13217)
> 
> Thank you for this issue ticket, I'll subscribe it.

That said, we can keep this issue open, in order to make dismissing these info messages survive reloads.

**Comment by MartinX3** [comment]
> Alright, we can store this information, and if it's dismissed one time, then one next session it won't appear again.

Thank you very much.

> I find this the wrong approach to deal with your issue If you want this unit to not start just disable it. systemctl disable podman. systemctl reset-failed. 

But i already disabled the root unit.
Since then this error is in the services tab of cockpit.

**Comment by KKoukiou** [comment]
@MartinX3 okay,maybe the cockpit podman page when it loads, it tries to start the system podman unit. If this fails to start you see it as failing unit in the services page. Even masking the unit wan't prevent this from appearing as failed. 
 
So these two issues are legit. Leaving it here and someone will take a look once there is time! 

Thanks for reporting.

**Comment by MartinX3** [comment]
@KKoukiou thank you very much :)

**Comment by martinpitt** [comment]
> Alright, we can store this information, and if it's dismissed one time, then one next session it won't appear again.

I'm not a fan of this, TBH. The user could never undo this and go back to system podman. What we could do is to not show the alert if the unit is masked, but there is a high chance that this is already the case?

**Comment by KKoukiou** [comment]
@martinpitt the user can always start the podman service from the 'Services' page.

**Comment by martinpitt** [comment]
Sure, it just feels like the place where you disable something should also be the place where to re-enable it. How would the user find the `podman.socket` unit even? It's rather easy to get confused about `podman.service`, which is *not* the thing to enable.

**Comment by KKoukiou** [comment]
@martinpitt I get your point, not having strong opinion here.
We could show a pop up dialog when a user closes the notification, which will provide info to the user on how to reenable it. However,  admittedly having a info notification in the page, is not that much of a nuisance to bother about such an implementation.

**Comment by marusak** [comment]
I agree with @martinpitt that we should not store this info but not show this message when the unit is masked.
At the same time I don't think that cockpit-podman needs to have this action supported. If someone really does not want to user system/user service, they can go and mask it. Then cockpit-podman should consider this unit as non-existent.

**Comment by garrett** [comment]
> This is desired and designed this way. It's an info notification and you can always dismiss it.

Actually, it's not, especially with the new podman design.

The banner just kind of showed up one day without any design input and it stuck around despite my opposition to it. :shrug: The language is fudgy too: "also available"... in addition to what exactly...?

Can't we check to see if there are podman containers on the system level and the user level and start the appropriate services to manage those containers? And then when creating a container, provide an option to create one for the system or for the user's session (and implicitly spin up the services needed)? 

Let's frame it this way:

1. A user wants to do something with containers
2. The user goes to the podman containers page
3. Instead of actually working with the containers they have (system or user session), they have to start up a daemon manually... every time they visit the page

No other part of Cockpit works like this, nor should it.

Visiting the podman page implicitly says "I want use podman".

> Sure, it just feels like the place where you disable something should also be the place where to re-enable it.

Yes. Fully agreed.

**Comment by martinpitt** [comment]
> Can't we check to see if there are podman containers on the system level

To do that, we first need the very podman.service that isn't currently running :grin: Agreed that the "also available" verbiage is confusing and should be improved. Maybe something like "Podman service for system|user containers is not running. [Start]"?

**Comment by garrett** [comment]
@martinpitt We should also have a remember option to ensure it always runs or is always ignored. The empty state that currently prompts starting the service should include both system and user and it should be remembered.

(This is all ignoring the root user, which is the same as system, basically, and shouldn't have a user/session service.)

If someone chooses to ignore a service, then it should be able to be re-enabled in services page, but that doesn't have support for user services (yet). So, therefore, to re-enable a user service, we'd need something like an options button in the podman page, similar to the one we have for the metrics.

So, a summary of what I think we're all suggesting together:

1. Change text for the popup
2. Handle user as well as system in the empty state
3. Have a remember option
4. Make an options button and an options modal for autostart of system, user, or system & user. (A none option would be strange, right? Or would it be "Always ask"?)

**Comment by garrett** [comment]
If the empty state is improved to handle both system and user _and_ we have the options button + modal dialog, then we could probably even skip having an alert on the page, so it'd really be this:

1. Handle user as well as system in the empty state, making sure remember works well here (where remember == `enable --now` and not remembering is only `start` without enable)
2. Make an options button and an options modal for autostart of system, user, or system & user
3. Remove the alert, depend on the empty state and podman's settings (which are really just a UI for the system and user `enable --now` and `disable` for the podman services)

**Comment by martinpitt** [comment]
Indeed: ideally, the way to "remember" is `enable --now`, and that's what the page is trying to do. This is more or less a fallback for [lots of](https://github.com/containers/podman/issues/8762) [podman](https://github.com/containers/podman/issues/9251) [bugs](https://github.com/containers/podman/issues/6660) (and it *still* happens) that make the API fail despite the service running.

Agreed that the text needs changing. WDYT about the proposal from above? "Podman service for system|user containers is not running. [Start]" (we use "Start" in libvirt as well, albeit with a service name: "Start libvirt"). 

Handling system and user together in one empty state instead of two alerts: enabling the API socket costs almost nothing thanks to socket activation, and as long as the API actually works, it might simplify the UI. My fear is mostly that it's still confusing to enable both, and then e.g. the user service crashes, as it would then still show system containers. So my gut feeling is that pretending that they are one service would aggravate the confusion.

**Comment by garrett** [comment]
> Agreed that the text needs changing. WDYT about the proposal from above?

Oh, I answered that already, but it looks like that was in the long reply that GitHub dropped somehow. :cry: I talked about having a remember checkbox and an ignore/start, IIRC. (We'll never know, as GitHub lost the comment.) But my 2 most recent comments since changed direction anyway, in an attempt to simplify things further.

In my most recent replies, I talked about getting rid of it entirely though. "Remove the alert, depend on the empty state and podman's settings"

> So my gut feeling is that pretending that they are one service would aggravate the confusion.

I never said anything about pretending they were one service though? I talked about having both being enable-able there in the empty state.

> enabling the API socket costs almost nothing thanks to socket activation

Could we just enable both by default then? Or do we need an opt-in?

(We could even possibly just forget the state management and always service start if they're not enabled or already running?)

> the user service crashes, as it would then still show system containers

Are we able to detect if the user service crashes? We shouldn't show "also available" when the service is intentionally enabled but isn't available. It should be an error message (with a solution to restart, naturally... or just attempt a restart ourselves and then show the error if the restart attempt doesn't work) if there's an error.

---

Basically, I'm suggesting this:

- Empty state (first run)
   - Include options for system _and/or user_ services to be enabled
- Main page
   - _Options_
      - _Enable/disable system and user services_
   - If a service isn't running, _but it's enabled_
      - _Attempt to run, but if that fails, then_ show an error
      - _Show an indication of some sort (with a spinner) that the service is being started and the page isn't ready to use yet_

I've italicized the new parts. 

I'm saying we should rid of the "also available" alert and depend on the empty state and options instead. We should check if the service is enabled or not. If there's an error where a service is not running (and is enabled), attempt to correct it and only show an error if we can't handle it (by trying to start it ourselves).

About the indication that services are starting and the page isn't ready yet: Having the page look empty when there are containers _or_ having the page look halfway full (and ready for action) when one of the services is not running are both wrong states. Either of these two states will cause a bad user experience. An artificial empty state or a partial state makes it look like containers were deleted or lost somehow.
